Trading Outpost now has half buy-only and half sell-only pallets

wafehling opened the pull request:
## About the PR
I added two new entity types, a buy and sell only cargo pallet, and added them to the trading outpost in place of the old pallets.

## Why / Balance
Since the day this was added, every single shift, at some point I'll hear someone in cargo yelling about how **"You sold stuff I just bought!"**

This seemed like the easiest way to fix the problem.

## Technical details
Added a variable to the cargo pallet component for labeling which type it is, added two new varieties of the cargo pallet to the yml, and adjusted all the calls for GetCargoPallets() to have boolean variables for buyOnly and sellOnly to tell the function what kind of call you're looking for. 

If you don't give it any specifics it'll still treat any pallet as a two-way one, but this function is only called 2 or 3 times and I've adjusted them all and commented the function itself for anyone looking to change/add it in the future. 

The old pallet is still in the game and is set to "both" by default, so it should work just fine for backwards compatibility. 

## Media
![image](https://github.com/space-wizards/space-station-14/assets/1471082/38cdf849-3e05-4f98-957f-65a719c2280d)
![image](https://github.com/space-wizards/space-station-14/assets/1471082/56e471e9-b45e-4332-bd55-7220566542ab)
![image](https://github.com/space-wizards/space-station-14/assets/1471082/a5e8e651-d301-420a-b2d7-cc4cf07a50af)
![image](https://github.com/space-wizards/space-station-14/assets/1471082/fdf2cf4c-fba8-46c9-9a30-7313a2204b5e)
![image](https://github.com/space-wizards/space-station-14/assets/1471082/3811edd0-bb32-44cf-b884-e0a9f4bf16fa)


- [X] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

## Breaking changes

Shouldn't break anything, I've left the old pallets fully in, and updated them to work with the new GetCargoPallets() function just like they used to. Not sure if/where you'd still need them, but better safe than sorry.

**Changelog**
:cl:
- tweak: The trading outpost now has dedicated buy-only and sell-only pallets. No more accidentally selling orders you just bought. Cargonians rejoice!

PolterTzi commented:
I feel like the spriting could use some work, but otherwise, yeah, good stuff.

wafehling commented:
> I feel like the spriting could use some work, but otherwise, yeah, good stuff.

I'm very much an amateur at it but I think it gets the point across at least. What would you tweak?

Interrobang01 commented:
Great change but the sprites are, as a maintainer would put it, horrendous. But that's what the review process is for.

Instead of crates, you might want to use the more culturally-familiar shopping cart. Or replace the closed crate with an open crate.
![image](https://github.com/space-wizards/space-station-14/assets/113810873/20fbfd3a-9cce-42ce-9c78-0aa2575bcc27)

Also, having the icon on every single pallet is way too noisy, so I would suggest making the icon low-contrast and putting a high-contrast icon as a decal on the trading post floor.

Alternatively just keep the old sprites because the pallets can be examined. Some color-coding would still be nice.

Keep accessibility in mind (colorblindness)

wafehling commented:
> Great change but the sprites are, as a maintainer would put it, horrendous. But that's what the review process is for.
> 
> Instead of crates, you might want to use the more culturally-familiar shopping cart. Or replace the closed crate with an open crate. ![image](https://private-user-images.githubusercontent.com/113810873/311480951-20fbfd3a-9cce-42ce-9c78-0aa2575bcc27.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTAwNDIwNjQsIm5iZiI6MTcxMDA0MTc2NCwicGF0aCI6Ii8xMTM4MTA4NzMvMzExNDgwOTUxLTIwZmJmZDNhLTljY2UtNDJjZS05Yzc4LTBhYTI1NzViY2MyNy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMzEwJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDMxMFQwMzM2MDRaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1hODI4YmY0ZDlkOTZkNmJjNTczZTlkZTk3YTAwMWRiOGY5Zjk2NDdkZjZhNjZiNTQ0ZjU3ZTUwZWRjYWFkZjA3JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.ZfSsxkkkGmGXDpogPgaqoy0NyFkELLB5LhtMeVpPyxw)
> 
> Also, having the icon on every single pallet is way too noisy, so I would suggest making the icon low-contrast and putting a high-contrast icon as a decal on the trading post floor.
> 
> Alternatively just keep the old sprites because the pallets can be examined. Some color-coding would still be nice.
> 
> Keep accessibility in mind (colorblindness)

Cargo isn't usually hauling shopping carts full of things around. I'll admit it's a bit bright and bold but c'mon.

I do agree on it being too noisy for each tile, Not sure how to do big decals but I'm sure I can figure it out. A decal of a big open crate with an arrow going in/out of it would work pretty well.

Tayrtahn commented on Content.Server/Cargo/Components/CargoPalletComponent.cs:
```diff
@@ -4,4 +4,11 @@ namespace Content.Server.Cargo.Components;
+    /// <summary>
+    /// Whether the pad is a buy pad, a sell pad, or both.
+    /// </summary>
+    [DataField("pallettype", required: false)]
```
```suggestion
    [DataField]
```
DataField infers the name from the property by default (you'll need to change capitalization elsewhere to "palletType").
I'm pretty sure required is false by default.

Tayrtahn commented on Content.Server/Cargo/Components/CargoPalletComponent.cs:
```diff
@@ -4,4 +4,11 @@ namespace Content.Server.Cargo.Components;
+    /// Whether the pad is a buy pad, a sell pad, or both.
+    /// </summary>
+    [DataField("pallettype", required: false)]
+    public string PalletType;
```
This should be an Enum instead of a string

wafehling replied:
I can't find a single example of someone using datafield with an enum anywhere, and it's complaining when I try. Are you sure about this

`    [DataField]
    public enum PalletType {buy, sell, both}`
Attribute 'DataField' is not valid on this declaration type. It is only valid on 'property, indexer, field' declarations. [Content.Server]

Tayrtahn replied:
Ah, not quite.

Define the enum type elsewhere (probably the bottom of the file)
```C#
public enum BuySellType
{
   Both,
   Buy,
   Sell
}
```

Then use that as the type for the DataField

```C#
[DataField]
public BuySellType PalletType;
```

If that doesn't work, let me know. You might need to specify a custom serializer for the DataField.

Tayrtahn commented on Content.Server/Cargo/Systems/CargoSystem.Shuttle.cs:
```diff
@@ -173,13 +173,16 @@ private void OnTradeSplit(EntityUid uid, TradeStationComponent component, ref Gr
-    private List<(EntityUid Entity, CargoPalletComponent Component, TransformComponent PalletXform)> GetCargoPallets(EntityUid gridUid)
+    /// GetCargoPallets(gridUid, true, false) to return only sell pads
+    /// GetCargoPallets(gridUid, false, true) to return only buy pads
+    private List<(EntityUid Entity, CargoPalletComponent Component, TransformComponent PalletXform)> GetCargoPallets(EntityUid gridUid, bool sellOnly = false, bool buyOnly = false)
```
This can just take a single Enum parameter and check if it matches the one on the component.

wafehling replied:
Alright I gotta be honest I have literally never worked with enums before and this shit is going to trip me up every single grueling step of the way. Can i leave it as is or is changing this one variable type crucial

Tayrtahn replied:
Just think of Enums as strings with type checking - if you make a typo, the compiler or your IDE will let you know right away. They're more efficient too.

So this would be
```C#
    private List<(EntityUid Entity, CargoPalletComponent Component, TransformComponent PalletXform)> GetCargoPallets(EntityUid gridUid, BuySellType palletType = BuySellType.Both)```

mirrorcult commented:
good pr

Tayrtahn commented on Resources/Prototypes/Entities/Objects/Specific/Cargo/cargo_pallet.yml:
```diff
@@ -53,3 +54,117 @@
+        shape:
+          !type:PhysShapeAabb
+          bounds: "-0.45,-0.45,0.45,0.45"
+        density: 15
+        mask:
+          - MachineMask
+  - type: CargoPallet
+    pallettype: "buy"
+  - type: StaticPrice
+    price: 100
+  - type: Sprite
+    drawdepth: FloorTiles
+    layers:
+    - sprite: Structures/cargo_pallets.rsi
+      state: cargo_pallet_buy
+  - type: Damageable
+    damageContainer: Inorganic
+    damageModifierSet: Metallic
+  - type: Destructible
+    thresholds:
+      - trigger:
+          !type:DamageTrigger
+          damage: 500
+        behaviors:
+          - !type:DoActsBehavior
+            acts: [ "Destruction" ]
+      - trigger:
+          !type:DamageTrigger
+          damage: 200
+        behaviors:
+          - !type:SpawnEntitiesBehavior
+            spawn:
+              PartRodMetal: # takes two to construct, so drop less than that
+                min: 0
+                max: 1
+          - !type:DoActsBehavior
+            acts: [ "Destruction" ]
+  - type: GuideHelp
+    guides:
+    - Cargo
```
Just set the base cargo pallet as parent and override the palletType

wafehling replied:
`- type: entity
  id: CargoPalletSell
  name: cargo selling pallet
  description: Designates valid items to sell with a selling computer, or to CentCom when a shuttle is recalled.
  parent: CargoPallet
  - type: CargoPallet
    palletType: "sell"
  - type: Sprite
    drawdepth: FloorTiles
    layers:
    - sprite: Structures/cargo_pallets.rsi
      state: cargo_pallet_sell
      
      Like this?`

Tayrtahn replied:
Almost! You still need the "components:" line. Sorry I didn't see you already had the parent specified in there.

When you use inheritance in entity definitions, the child will automatically get all of the components that the parent has, with the same values for all the properties. You should only list a component type again in the child if you're changing some of its values. That way, any future changes to the parent will automatically be applied to the child too

wafehling replied:
That makes sense. Updated that in the yml, should be good to go here.

Tayrtahn reviewed (changes requested):
There are a few things that need some cleanup, but I like this overall

beck-thompson commented:
Honestly I think they look fine! I'm no expert so I wont make suggestions but overall 👍 👍

TheShuEd commented:
![in](https://github.com/space-wizards/space-station-14/assets/96445749/f96aec22-47b3-4c99-a8a3-685b4f2292b1)
![out](https://github.com/space-wizards/space-station-14/assets/96445749/75e53cc9-a550-48cc-bf6d-fa92b6790283)

metalgearsloth commented on Resources/Textures/Structures/cargo_pallets.rsi/meta.json:
```diff
@@ -0,0 +1,17 @@
+{
+    "version": 1,
+    "license": "CC-BY-SA-3.0",
+  "copyright": "Created by wafehling (github) partially from a modified creation by EmoGarbage404 (https://github.com/space-wizards/space-station-14/blob/e3180aec030edc6fef25e0262831c2215d9bb2a8/Resources/Textures/Structures/Storage/Crates/generic.rsi/meta.json)",
+    "size": {
+        "x": 32,
+        "y": 32
+    },
+    "states": [
+        {
+             "name": "cargo_pallet_buy"
+        },
+        {
+            "name": "cargo_pallet_sell"
+        }
+    ]
+}
```
You should run a formatter over this as it's a bit skewiff.

wafehling replied:
Fixed a couple scraggly spaces, thanks.

wafehling commented:
> I quickly drew sprites, which in my opinion would be better suited ![in](https://private-user-images.githubusercontent.com/96445749/311507876-f96aec22-47b3-4c99-a8a3-685b4f2292b1.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTAwODUxOTgsIm5iZiI6MTcxMDA4NDg5OCwicGF0aCI6Ii85NjQ0NTc0OS8zMTE1MDc4NzYtZjk2YWVjMjItNDdiMy00Yzk5LWE4YTMtNjg1YjRmMjI5MmIxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMTAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzEwVDE1MzQ1OFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWU0N2ViMzc1ODZkNzlhNzQ4MzNkZTkzMWYwNWE1MDUxY2E3NjcwYWRiZWI0YjRmZjUzYjBlZWFlNDI0NzllMGImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.nO_BhiuL3vCG6_G59m6R6HXkLC9lgmajQQ-MLlmJghA) ![out](https://private-user-images.githubusercontent.com/96445749/311507882-75e53cc9-a550-48cc-bf6d-fa92b6790283.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTAwODUxOTgsIm5iZiI6MTcxMDA4NDg5OCwicGF0aCI6Ii85NjQ0NTc0OS8zMTE1MDc4ODItNzVlNTNjYzktYTU1MC00OGNjLWJmNmQtZmE5MmI2NzkwMjgzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMTAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzEwVDE1MzQ1OFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTM2MTBjMGZlMDUyZTdhOTIwYmVjMDZlNGE4NTAyODAwMDgzMjRlODRlYTAzNzJmNGFhOGZkMzYyMTY3ZDI2OWYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.H_2bi5btfBue_g4B8YO7wjpKaDYKgliPChsKsEtgfXA)

Oh I REALLY like that. The little arrows in the corner takes care of the colorblind issue too. I can integrate those and credit you if you'd like?

TheShuEd commented:
yeah, sure

Tayrtahn commented on Content.Server/Cargo/Systems/CargoSystem.Shuttle.cs:
```diff
@@ -190,7 +193,17 @@ private List<(EntityUid Entity, CargoPalletComponent Component, TransformCompone
                 continue;
             }
 
+            if (sellOnly && comp.PalletType != "sell" && comp.PalletType != "both")
+            {
+                continue;
+            }
+            if (buyOnly && comp.PalletType != "buy" && comp.PalletType != "both")
+            {
+                continue;
+            }
+
```
And down here you can just do something like

```C#
if (palletType != comp.PalletType && comp.PalletType != PalletType.Both)
   continue;
```
Which will cover both cases.

(I'm writing this on my phone, so you may need to modify things a bit to make it work)

Tayrtahn commented on Content.Server/Cargo/Systems/CargoSystem.Shuttle.cs:
```diff
@@ -250,7 +263,7 @@ private void GetPalletGoods(EntityUid gridUid, out HashSet<EntityUid> toSell, ou
         toSell = new HashSet<EntityUid>();
 
-        foreach (var (palletUid, _, _) in GetCargoPallets(gridUid))
+        foreach (var (palletUid, _, _) in GetCargoPallets(gridUid, true, false))
```
Then this would be
`GetCargoPallets(gridUid, BuySellType.Sell)`

wafehling commented:
Alrighty! I updated the textures and the meta credit, and changed the checker from using a string to using an enum. Fixed the meta janky spacing as well.

Went back in and tested it and we've got it working flawlessly. I can't think of anything else that needs fixing.

![image](https://github.com/space-wizards/space-station-14/assets/1471082/9246788f-5665-452c-9ff6-b40e765a7f51)
![image](https://github.com/space-wizards/space-station-14/assets/1471082/05e3b465-cf7b-4bac-9d7f-9f7a7b1cae3c)
![image](https://github.com/space-wizards/space-station-14/assets/1471082/bd422b2b-90ee-4347-89e5-4a169c77031b)
![image](https://github.com/space-wizards/space-station-14/assets/1471082/975316e7-bd0e-45ba-b27f-37b8a0b7fd9b)
![image](https://github.com/space-wizards/space-station-14/assets/1471082/5b7e9cf8-d956-48cd-90d8-a354a188e968)

TheShuEd commented:
Maybe add icon smoothing? 🤔

UbaserB commented:
> Maybe add icon smoothing? 🤔

i kind of like it without

wafehling commented:
> Maybe add icon smoothing? 🤔

Man you made the icons

Can we please just merge this someone sold my steel crates today

How did this trade station get approved like this and this is under so much scrutiny lol

Tayrtahn commented:
That's the fun of PRs: you never know if you're going to get a blind merge or excruciating nitpicking!

I think they look fine as-is, but I'd be curious to see how they look with icon smoothing. Maybe someone familiar with how to set that up could try it out as a follow-up PR?

EmoGarbage404 commented:
The sprite looks fine as is.

EmoGarbage404 commented:
this is a 3 day old pr that got a first round of review literally two days ago. there are 250-odd PRs on the repo that need reviewing and there's always more. There's no need to sit around begging for merges. it's gonna get reviewed again once someone has time.

metalgearsloth commented on Content.Server/Cargo/Components/CargoPalletComponent.cs:
```diff
@@ -1,7 +1,24 @@
-public sealed partial class CargoPalletComponent : Component {}
+public sealed partial class CargoPalletComponent : Component
+{
+    /// <summary>
+    /// Whether the pad is a buy pad, a sell pad, or both.
+    /// </summary>
+
+    public enum BuySellType
+    {
+        buy,
+        sell,
+        both
+    }
+
```
C# style guide should be followed, enums should be pascal case.

Also just make this flags and then have All = Buy | Sell;

metalgearsloth commented on Content.Server/Cargo/Systems/CargoSystem.Shuttle.cs:
```diff
@@ -8,6 +8,7 @@
 using Robust.Shared.Map;
 using Robust.Shared.Random;
 using Robust.Shared.Audio;
+using BuySellType = Content.Server.Cargo.Components.CargoPalletComponent.BuySellType;
 
```
Just put the enum outside of the class so you don't get these clunky imports.

metalgearsloth commented on Content.Server/Cargo/Systems/CargoSystem.Shuttle.cs:
```diff
@@ -190,7 +194,17 @@ private List<(EntityUid Entity, CargoPalletComponent Component, TransformCompone
                 continue;
             }
 
+            if (requestType == BuySellType.sell && comp.PalletType != BuySellType.sell && comp.PalletType != BuySellType.both)
+            {
+                continue;
+            }
+            if (requestType == BuySellType.buy && comp.PalletType != BuySellType.buy && comp.PalletType != BuySellType.both)
+            {
+                continue;
+            }
+
```
Just use bitmasks for this (see stuff like (X & Y) != 0x0)

wafehling replied:
I have no idea how to do any of this stuff, my dude

wafehling replied:
You're gonna have to elaborate on this one.

wafehling commented:
Changed enum to flags and bitmask-friendly format,
Moved enum outside of the class,
Used bitmasks for more compact if checks

Done, done, and done. 

Tested and working.

metalgearsloth commented on Content.Server/Cargo/Components/CargoPalletComponent.cs:
```diff
@@ -1,7 +1,25 @@
 /// <summary>
 /// Any entities intersecting when a shuttle is recalled will be sold.
 /// </summary>
+
+public enum BuySellType
+{
+    Buy = 1,
+    Sell = 2,
+    All = Buy | Sell
+}
+
```
```suggestion

[Flags]
public enum BuySellType : byte
{
    Buy = 1 << 0,
    Sell = 1 << 1,
    All = Buy | Sell
}

```

metalgearsloth commented on Content.Server/Cargo/Systems/CargoSystem.Shuttle.cs:
```diff
@@ -190,7 +193,14 @@ private List<(EntityUid Entity, CargoPalletComponent Component, TransformCompone
                 continue;
             }
 
+            if (((requestType & BuySellType.Sell) != 0 && (comp.PalletType & BuySellType.Sell) == 0) ||
+                ((requestType & BuySellType.Buy) != 0 && (comp.PalletType & BuySellType.Buy) == 0))
+            {
```
```suggestion
            if ((requestType & comp.PalletType) == 0)
            {
```

wafehling replied:
That... makes a lot more sense, yeah. Ha.
